src/workerd/server/workerd.capnp

```diff
@@ -601,6 +601,40 @@ struct Worker {
 
   moduleFallback @13 :Text;
 
+  struct Limits {
```

Comment:
I'm concerned that this limits structure offers too much configurability, which has two problems:
* We'll be obliged to support it forever, which may be difficult if the way heap enforcement works in V8 changes. Like, what if they deprecate the near-heap-limit callback and it's no longer possible to apply a multiplier to the limit there? What if the GC algorithm changes and "young generation size" is no longer an option? (In fact, when the workers runtime was first written, the GC algorithm was different and had entirely different options. It can change again.)
* It's probably confusing for developers who don't actually want to control things in so much detail. Why would someone want to change the multiplier or the number of times it is applied? Who would ever want to tune the young generation size?

Can we make this more abstract? I'd suggest two settings:

* The heap limit.
* A boolean or enum that chooses between warn and terminate.
* A boolean or enum that chooses whether to dump a heap snapshot.

Reply:
I'd suggest if the user chooses "warn" mode, when the near-heap-limit callback runs and writes the warning, it should then change the limit to unlimited.

Reply:
To head off the obvious question:

"What if people want to emulate our production behavior exactly, where an isolate that has gone over the limit is permitted to complete in-flight requests, as long as it doesn't go excessively over?"

I don't think that's a behavior people should be testing against. This behavior is designed to be lenient to code that is already deployed. But when testing, developers should enforce the limit strictly, and design their app so it never exceeds that limit in the first place. Moreover, we might change our production behavior in the future, probably making it a bit more strict, but probably in a nuanced way.

Reply:
Can you explain the use case for allowing the heap limit to be exceeded multiple times?

Reply:
> Can you explain the use case for allowing the heap limit to be exceeded multiple times?

For memory leak analysis I want to be able to generate 2 or more heap-snaphots at various points in order to generate a comparison view. In most cases I likely only need 2 (which is why the defaults are set the way they are here now) but generating a third or fourth snapshot can be help in some cases.

Reply:
Oh I see. But is generating these on near-heap-limit callbacks actually the best approach? I've more commonly seen time-based, that is, take a snapshot, wait 10 seconds, take another one. It might also be nice if the snapshots can be taken between requests, to reduce the amount of temporary garbage that shows up in them...

What about having an API exposed to JavaScript for taking a snapshot, similar to `--expose-gc`?

Reply:
For time-based or snapshots between requests, using the devtools debugger is likely still the better approach or we can add an API for it similar to what Node.js provides.

I can see us allowing for multiple approaches here.

For reference to what Node.js allows...

* nearlimit snapshot via cli option: https://nodejs.org/docs/latest/api/cli.html#--heapsnapshot-near-heap-limitmax_count
* Generate snapshot when signal received: https://nodejs.org/docs/latest/api/cli.html#--heapsnapshot-signalsignal
* API for taking snapshots - https://nodejs.org/docs/latest/api/v8.html#v8writeheapsnapshotfilenameoptions
* API for setting nearlimit snapshot: https://nodejs.org/docs/latest/api/v8.html#v8setheapsnapshotnearheaplimitlimit

Reply:
For now, let's defer discussion of a possible API-based approach for generating snapshots to a separate PR.

I have no particularly strong opinions on exactly what the config options here should be beyond making it possible to define some reasonable threshold and getting 1+ heap snapshots generated automatically when we hit the threshold. So I'm open to specific suggestions on which set of config options to use.